### PR TITLE
lib/*-plugins: initial support for `lib.mkPackageOption`

### PR DESCRIPTION
**Pull request by MattSturgeon**
## Summary

Adds initial support for receiving a package _option_ instead of a package _derivation_ in `mkVimPlugin` and `mkNeovimPlugin`. This is done to enable the use of [`lib.mkPackageOption`](https://nixos.org/manual/nixos/unstable/#sec-option-declarations-util-mkPackageOption) as discussed in #1950.

This soft-deprecates our current helpers, `lib.nixvim.mkPluginPackageOption` and `lib.nixvim.mkPackageOption`.

Also implemented this for the plugin [TEMPLATE](https://github.com/nix-community/nixvim/pull/2139/files#diff-93af850d88bc329f8ac9a11854624cfd9cf99d723bac17464f4dac5993425f3d), [cmp sources](https://github.com/nix-community/nixvim/pull/2139/files#diff-961bb62e8020d29c7e320ed7a45345f1fa86e338342bb44df02bd161e7ff96c2), telescope extensions, and [lsp language-servers](https://github.com/nix-community/nixvim/pull/2139/files#diff-84e25e99a6dafddf43e738af5502df469f8deb4f4eed194544c2de10133abe00).

## Why

An option with a package default, will be rendered using `lib.generators.toPretty` as (e.g.) `<derivation vimplugin-cmp-ai-2024-08-01>`.

We could specify a `defaultText` for all our package options, e.g. `defaultText = lib.literalExpression "pkgs.cmp-ai"`, however `lib.mkPackageOption` automates this by setting both `default` and `defaultText` based on an attr-path.

#### Screenshots

<details><summary>Before+after screenshots</summary>
<p>

### Before

image

### After

image

</p>
</details>

## Alternative/future approach

It would be **possible** to have `mk*Plugin` handle calling `lib.mkPackageOption` internally, however it'd have to do so within an imported module in order to gain access to the (correct) `pkgs` arg[^1].

We may need to do this _anyway_ to also gain access to `options` and `config` without having to pass them into the helpers, but that is out of scope for this PR...

See this POC:

<details><summary>Proof of concept</summary>
<p>

```nix
mkNeovimPlugin =
  {
    name,
    defaultPackage,
    # omitted for clarity
  }:
  {
    imports = [
      # Import a module, so we have access to the "correct" pkgs
      ({ config, options, pkgs, ... }: {
        options.plugins.${name} = {
          enable = lib.mkEnableOption name;
          package =
            if lib.isOption defaultPackage then
              defaultPackage
            else
              lib.mkPackageOption pkgs name {
                default = defaultPackage;
              };
        };
      })
    ];
    # Rest omitted for clarity
  }
```

</p>
</details> 

[^1]: As part of `helpers`, we do have access to _a_ `pkgs`, but it may not be the correct one configured for use within the module eval. Incidentally, this is making me think we should separate out all helper functions that rely on `pkgs` and put them somewhere else so that `helpers` doesn't need a `pkgs`.

### CODE REVIEW

File: lib/vim-plugin.nix [hunk 1]
```diff
@@ -5,7 +5,7 @@ with lib;
     config:
     {
       name,
-      url ? if defaultPackage != null then defaultPackage.meta.homepage else null,
+      url ? if packageOption == null then null else packageOption.default.meta.homepage,
```
Comment (traxys) [inline lib/vim-plugin.nix:8]: nit: can't this be done with a `or` ?

**Review by traxys — COMMENTED** [review]
It would have been nice to remove completly the `defaultPackage` in this PR, it would simplify a bit the helpers, though I think I saw there are around 150 occurences, it may be some work...

Maybe a regex could replace them all? Not sure

File: lib/vim-plugin.nix (hunk 1 above)
Comment (MattSturgeon) [reply lib/vim-plugin.nix:8]: Actually this isn't needed; when I first started work on this `defaultPackage` and `packageOption` were nullable and the `package` option would be omitted from `options`.

That behaviour is never used in nixvim, so I dropped it; `packageOption` is no longer nullable.

To answer your original question, using `or` would cause `packageOption.default.meta.homepage` to silently be null when the package has no homepage, however we'd want that to error so the maintainer knows he has to set `url` explicitly.

**Comment by MattSturgeon** [comment]
> It would have been nice to remove completely the `defaultPackage` in this PR, though I think I saw there are around 150 occurences, it may be some work...

I agree it'd be nice, however I don't want to rush to replace all `defaultPackage`s with `lib.mkPackageOption` if we're still considering the imported-module approach.

A treewide substitution of `defaultPackage = pkgs.foo` -> `defaultPackage = "foo"` would be much easier as well...

**Comment by traxys** [comment]
Indeed, it would be nice to leave the `defaultPackage` argument when no customization is required

**Comment by khaneliman** [comment]
So this gives us clearer documentation and nullable packages for plugins, which are big wins to resolve https://github.com/nix-community/nixvim/issues/1970 and https://github.com/nix-community/nixvim/issues/1950. Does it also have the benefit of the lazy package evaluation to prevent pkg errors being thrown from unused options or is that still an issue that requires a a different solution?

**Comment by traxys** [comment]
I think because we include the man page by default it will never be lazy, whatever we do. We may need to reconsider the man page inclusion by default

**Comment by MattSturgeon** [comment]
I have some gripes with how we build the man docs too; they will always use our flake's pkgs rather than the pkgs used in module eval.

To achieve laziness, I think we need to either ditch the man pages being included by default, and/or remove automatically getting the `meta.homepage` from the default package.

**Comment by khaneliman** [comment]
Did you wanna add a test to verify how it responds to null packages with the new functionality?

**Comment by MattSturgeon** [comment]
We already have some `null` packages in lsp/language-servers, and upstream document how it should behave when `default = null`, so I don't see it as a major issue. If you have a suggestion for a test case I'm happy to consider adding something though.

**Comment by khaneliman** [comment]
Sorry, was in the mindset of top level package option. These were already nullable.

File: lib/neovim-plugin.nix [hunk 2]
```diff
@@ -29,7 +29,12 @@ with lib;
       colorscheme ? name,
       # options
       originalName ? name,
-      defaultPackage,
+      # WARNING: `defaultPackage` is being replaced with `packageOption`,
+      # which should be defined using: lib.mkPackageOption pkgs originalName { default = [ "path" "to" "package" ]; }
+      # NOTE: We can't do that here (without importing a module) because we don't have access to the `pkgs` module arg.
```
Comment (MattSturgeon) [inline lib/neovim-plugin.nix:34]: Should this be tackled in this PR? I.e. implementing the package option within an imported-module

Comment (traxys) [reply lib/neovim-plugin.nix:34]: It would not need to much additional complexity with #2140 right?

**Comment by MattSturgeon** [comment]
> So this gives us clearer documentation and nullable packages for plugins

@khaneliman just to clarify, this does **not** introduce nullable package options, unless a maintainer chooses to explicitly pass `nullable = true` to `mkPackageOption`, like I did in mkLsp.

**Comment by MattSturgeon** [comment]
@traxys this now migrates all call-sites for `mkVimPlugin` and `mkNeovimPlugin` to use the new pattern, so I've removed the old `defaultPackage` argument from those functions.

I don't want to remove `helpers.mkPackageOption` and `helpers.mkPluginPackageOption` in _this_ PR, because the treewide parts are already getting large, and will be difficult to review if we also have to check various ad-hoc package options are migrated correctly (e.g. options like `plugins.*.gitPackage`).

There's currently 89 matches for `mkPluginPackageOption` and 60 matches for `mkPackageOption` in the repo, although some of those are for `lib.mkPackageOption` (there's 40 matches for `/helpers\.mkPackageOption/`).

File: docs/user-guide/faq.md [hunk 3]
```diff
@@ -29,6 +29,8 @@ The [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#managing-plugins-w
 
 When using NixVim it is possible to encounter an error of the type `attribute 'name' missing`, for example it could look like:
 
+<!-- TODO: Update example now that we use `mkPackageOption` -->
```
Comment (khaneliman) [inline docs/user-guide/faq.md:32]: Did we want to handle this in this PR or follow up?

**Review by khaneliman — APPROVED** [review]
LGTM, I very much appreciate the direction of this PR.

File: docs/user-guide/faq.md (hunk 3 above)
Comment (MattSturgeon) [reply docs/user-guide/faq.md:32]: I think we'll leave it for later. Next time a user reports a mismatched nixpkgs error we can use the error message they report in the example here.

**Comment by MattSturgeon** [comment]
@mergifyio queue
